openmc/data/endf.py

```diff
@@ -0,0 +1,409 @@
+"""Module for parsing and manipulating data from ENDF evaluations.
+
+All the classes and functions in this module are based on document
+ENDF-102 titled "Data Formats and Procedures for the Evaluated Nuclear
+Data File ENDF-6". The latest version from June 2009 can be found at
+http://www-nds.iaea.org/ndspub/documents/endf/endf102/endf102.pdf
+
+"""
+from __future__ import print_function, division, unicode_literals
+
+import io
+import re
+import os
+from math import pi
+from collections import OrderedDict, Iterable
+
+import numpy as np
+from numpy.polynomial.polynomial import Polynomial
+
+from .function import Tabulated1D, INTERPOLATION_SCHEME
+from openmc.stats.univariate import Uniform, Tabular, Legendre
+
+
+LIBRARIES = {0: 'ENDF/B', 1: 'ENDF/A', 2: 'JEFF', 3: 'EFF',
+             4: 'ENDF/B High Energy', 5: 'CENDL', 6: 'JENDL',
+             31: 'INDL/V', 32: 'INDL/A', 33: 'FENDL', 34: 'IRDF',
+             35: 'BROND', 36: 'INGDB-90', 37: 'FENDL/A', 41: 'BROND'}
+
+SUM_RULES = {1: [2, 3],
+             3: [4, 5, 11, 16, 17, 22, 23, 24, 25, 28, 29, 30, 32, 33, 34, 35,
+                 36, 37, 41, 42, 44, 45, 152, 153, 154, 156, 157, 158, 159, 160,
+                 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172,
+                 173, 174, 175, 176, 177, 178, 179, 180, 181, 183, 184, 185,
+                 186, 187, 188, 189, 190, 194, 195, 196, 198, 199, 200],
+             4: list(range(50, 92)),
+             16: list(range(875, 892)),
+             18: [19, 20, 21, 38],
+             27: [8, 101],
+             101: [102, 103, 104, 105, 106, 107, 108, 109, 111, 112, 113, 114,
+                   115, 116, 117, 155, 182, 191, 192, 193, 197],
+             103: list(range(600, 650)),
+             104: list(range(650, 700)),
+             105: list(range(700, 750)),
+             106: list(range(750, 800)),
+             107: list(range(800, 850))}
+
+_ENDF_FLOAT_RE = re.compile(r'([\s\-\+]?\d*\.\d+)([\+\-]\d+)')
+
+
+def radiation_type(value):
+    p = {0: 'gamma', 1: 'beta-', 2: 'ec/beta+', 3: 'IT',
+         4: 'alpha', 5: 'neutron', 6: 'sf', 7: 'proton',
+         8: 'e-', 9: 'xray', 10: 'unknown'}
+    if value % 1.0 == 0:
+        return p[int(value)]
+    else:
+        return (p[int(value)], p[int(10*value % 10)])
+
+
+def float_endf(s):
+    """Convert string of floating point number in ENDF to float.
+
+    The ENDF-6 format uses an 'e-less' floating point number format,
+    e.g. -1.23481+10. Trying to convert using the float built-in won't work
+    because of the lack of an 'e'. This function allows such strings to be
+    converted while still allowing numbers that are not in exponential notation
+    to be converted as well.
+
+    Parameters
+    ----------
+    s : str
+        Floating-point number from an ENDF file
+
+    Returns
+    -------
+    float
+        The number
+
+    """
+    return float(_ENDF_FLOAT_RE.sub(r'\1e\2', s))
```

Comment:
Yes, the regex has to be "compiled" before use, so declaring it module-level allows the compilation to happen only once which in theory would reduce the time for repeated calls to this function. After doing a little reading, apparently Python actually [caches the compilation anyway](http://stackoverflow.com/questions/452104/is-it-worth-using-pythons-re-compile), so perhaps there's not such a difference in performance.